File: fib-mem.js
const fib = function (n, cache) {
  if (n <= 1) {
    return 1
  }
  let a = cache[n - 1]
  let b = cache[n - 2]
  if (!a) {
    a = fib(n - 1, cache)
    cache[n - 1] = a
  }
  if (!b) {
    b = fib(n - 2, cache)
    cache[n - 2] = b
  }
  return a + b
}

const cache = new Array(46)
console.log(fib(46, cache))

File: fib-mem.swift

func fib(_ n: Int, cache: inout [Int:Int]) -> Int {
    if n <= 1 { return 1 }

    if let f = cache[n] {
        return f
    }

    let a = fib(n - 1, cache: &cache)
    let b = fib(n - 2, cache: &cache)

    let f = a + b
    cache[n] = f
    return f
}

var cache = [Int:Int]()
print(fib(46, cache: &cache ))
